src/constructors/lattices.c

```diff
@@ -305,3 +314,287 @@ igraph_error_t igraph_triangular_lattice(
 
     return IGRAPH_SUCCESS;
 }
+
+
+/**
+ * Creates a hexagonal lattice whose vertices have the form (i, j) for non-negative integers i and j
+ * and (i, j) is connected with (i + 1, j), and if i is odd also with (i - 1, j + 1) provided a vertex
+ * exists. Thus, all vertices have degree at most 3.
+ *
+ * </para><para>
+ * The vertices of the resulting graph are ordered lexicographically with the 2nd coordinate being
+ * more significant, e.g., (i, j) &lt; (i + 1, j) and (i + 1, j) &lt; (i, j + 1).
+ *
+ * \param graph An uninitialized graph object.
+ * \param directed Boolean, whether to create a directed graph.
+ *        If the \c mutual argument is not set to true,
+ *        edges will be directed from lower-index vertices towards
+ *        higher-index ones.
+ * \param mutual Boolean, if the graph is directed this gives whether
+ *        to create all connections as mutual.
+ * \param row_lengths_vector Integer vector, defines the number of vertices with
+ *        the second coordinate equal to the index. The length of this vector must match
+ *        the length of \p row_start_vector. All coordinates must be non-negative.
+ * \param row_start_vector Integer vector, defines the leftmost coordinate of
+ *        the vertex with the second coordinate equal to the index.
+ *
+ * \return Error code:
+ *         \c IGRAPH_EINVAL: invalid (negative) length of row_lengths_vector does not match the length of the
+ *         row_start_vector.
+ *
+ * Time complexity:  O(|V|), where |V| is the number of vertices in the generated graph.
+ */
+static igraph_error_t hexagonal_lattice(
+    igraph_t *graph, igraph_bool_t directed, igraph_bool_t mutual,
+    const igraph_vector_int_t *row_lengths_vector, const igraph_vector_int_t *row_start_vector
+) {
+    igraph_vector_int_t edges = IGRAPH_VECTOR_NULL;
+    igraph_integer_t row_count = igraph_vector_int_size(row_lengths_vector);
+    igraph_integer_t no_of_nodes;
+    igraph_vector_int_t row_lengths_prefix_sum_vector;
+    igraph_integer_t i, j;
+    igraph_bool_t lex_ordering = false;
+
+    if (igraph_vector_int_size(row_lengths_vector) != igraph_vector_int_size(row_start_vector)) {
+        IGRAPH_ERRORF(
+            "Length of row_lengths_vector vector (%" IGRAPH_PRId ") must match the length of the "
+            "row_start_vector (%" IGRAPH_PRId ").",
+            IGRAPH_EINVAL,
+            igraph_vector_int_size(row_lengths_vector),
+            igraph_vector_int_size(row_start_vector)
+        );
+    }
+
+    for (i = 0; i < row_count; i++) {
+        if (VECTOR(*row_lengths_vector)[i] < 0) {
+            IGRAPH_ERRORF(
+                "row_lengths_vector vector must have non-negative coordinates, "
+                "was (%" IGRAPH_PRId ") for the (%" IGRAPH_PRId ")-th row.",
+                IGRAPH_EINVAL, VECTOR(*row_lengths_vector)[i], i);
+        }
+    }
+
+    IGRAPH_VECTOR_INT_INIT_FINALLY(&edges, 0);
+
+    COMPUTE_NUMBER_OF_VERTICES
```

Comment:
Instead of just using `{ ... }`, please use `do { .... } while (0)` when defining the macro. This will shut up some linters that would complain about an empty statement when the semicolon is added.

Actually, I'm wondering if this is better converted a real function ... but that's not really important.

Reply:
Let me know if it is Ok now.